Refactor to decouple classes

chahak13 opened the pull request.

chahak13 committed:
Add example for 1d bar

chahak13 committed:
Implement shapefn as ABC and separate 1d/2d class

chahak13 committed:
Start adding 2d mesh functions

chahak13 committed:
Add func to update particle element mapping in 2d

chahak13 committed:
Update and test more functions for 2D

chahak13 committed:
Update methods for 2d and add tests

chahak13 committed:
Update force functions for 2D.

chahak13 committed:
Running version of 2d solve

Correctness to be checked.

chahak13 committed:
Refactor Particle class

chahak13 committed:
Refactor Material class

chahak13 committed:
Add Element containers

chahak13 committed:
Add more methods to element, particles and node

chahak13 committed:
Add more functions

chahak13 committed:
Add details for Linear Elastic material

chahak13 committed:
Functions to calculate force etc

chahak13 committed:
Add scheme and solver classes

kks32 commented on diffmpm/element.py:
```diff
@@ -1,30 +1,539 @@
+
+    def id_to_node_loc(self, id: int):
+        """
+        Node locations corresponding to element `id`.
```
Is this the physical location of a node? then shouldn't that depend on the element size?

chahak13 replied:
I set the locations by multiplying with the element size so that `loc` are the coordinates and not just the ids. Is that not how it should be?

kks32 replied:
That works...(for most cases). So I won't worry about it. In MPM, most often the mesh has the same element size throughout.

SachinJalan commented on diffmpm/element.py:
```diff
@@ -1,30 +1,539 @@
+        Arguments
+        ---------
+        xi : float, array_like
+            Locations of particles in natural coordinates to evaluate
```
natural coordinates specify which coordinates, the background mesh ones with nodes or the material points one.

chahak13 replied:
Natural coordinates are coordinates in the element, relative to the node. For example, if we're talking about 1D elements, say the mesh looks something like this:
```
+--------+-------+----o----+-------+-------+
0        1       2    p    3       4       5
```
Here, if p is the point right in the center of the element with nodes 2 and 3, then it's physical coordinates would be x=2.5. Now when the element is mapped to a standard element
```
 +----o----+
-1    p    1
```
Then it's natural coordinate will be x=0. Does that make sense?

SachinJalan replied:
yes, now its clear

chahak13 committed:
Add basic 1D working setup

chahak13 commented:
A quick update. With the current setup, the 1d solver "works", i.e. it can solve a 1d axial bar problem without JIT. For JIT, the problem is slightly non-trivial now because we have more classes that move around than before. To make a class JIT-compatible, it needs the `tree_flatten` and `tree_unflatten` methods (which will then be used to register the class as a PyTree node). While this is relatively easy, in our case, when we instantiate the objects, we also initialize extra arrays for various things (velocity, force etc.). Now, having this in `__init__()` doesn't work as during JIT it "reinitializes" the objects and hence resets everything. The way I got around this issue previously was by passing these variables in the constructor and initializing only when passed values are `None` but this is not very clean. Since I know that method works, I'm going to first try having the initialization of arrays as another method that the solver calls and doesn't happen on instantiation (a bit like what's done in cb-geo/mpm) and see if I can make it workable with JIT. If not, will have to go with the previous method. Doing more reading to see if there are any other ways

chahak13 commented:
A problem with the separate initialization method is that those attributes are lost when unflattened. Checked a few different code bases on github that seem to be doing something similar for some variables so will have to go with the case of having all of them in `__init__`
1. https://github.com/ott-jax/ott/blob/64d26ac4bb85a732d5e9d11010c9ae1394a99a6c/src/ott/initializers/nn/initializers.py#L35
2. https://github.com/isabella232/ott/blob/56a9922f26b0e44be4a2d3de46808458cfe3e8cc/ott/core/sinkhorn.py#L268
3. https://github.com/lindermanlab/ssm-jax/blob/3ee7580d0901e10ac5e237655e81133d5e2a67ca/ssm/hmm/transitions.py#L5

chahak13 committed:
Add pytree methods to support JIT

chahak13 commented:
We have JIT support now. Here's an example for optimization with JIT
![image](https://github.com/geoelements/diff-mpm/assets/29800965/79ad519b-bcde-4fea-b316-43360a4d6cf8)

chahak13 committed:
Correct damping energy issue

The nodal velocity update should be using the particle velocity values
mapped onto the nodes, which was not happening earlier. This led to the
total energy of the system being damped over time. This is now corrected.

chahak13 committed:
Update example and store more solver results

chahak13 commented:
Any suggestions on how to select which values to return from the solver? Should it be user dependent or should it just return all of it i.e. position, velocity, stress, strain, energies etc.? Trying to think what would be a good user experience. Also, should I add 2D to this PR itself or once this refactoring is approved, merge this and create new for 2d?

kks32 commented:
Let's do a separate 2D implementation, we can approve this PR

chahak13 commented:
Okay, sounds good, thanks. Let me clean it up a little and I'll move it out of draft.

chahak13 committed:
Code cleanup; doesn't change any functionality

chahak13 committed:
Swap locations of two functions

`set_particle_element_ids` is moved to element.py since the element id
mapping is based on what type of element the particles lie in. On the
other hand, `update_natural_coords` was moved to particle.py since it is
an intrinsic property of the particles and is just updated based on the
elements passed as argument.

chahak13 commented:
@kks32 this should be good to review now, I think. Let me know if you have any thoughts or feel free to merge otherwise. Thanks!

chahak13 committed:
Add simple version of file input and parser
